homeassistant/components/cert_expiry/config_flow.py

```diff
@@ -34,69 +24,59 @@ def __init__(self) -> None:
         """Initialize the config flow."""
         self._errors = {}
 
-    def _prt_in_configuration_exists(self, user_input) -> bool:
-        """Return True if host, port combination exists in configuration."""
-        host = user_input[CONF_HOST]
-        port = user_input.get(CONF_PORT, DEFAULT_PORT)
-        if (host, port) in certexpiry_entries(self.hass):
-            return True
-        return False
-
     async def _test_connection(self, user_input=None):
-        """Test connection to the server and try to get the certtificate."""
-        host = user_input[CONF_HOST]
+        """Test connection to the server and try to get the certificate."""
         try:
-            await self.hass.async_add_executor_job(
-                get_cert, host, user_input.get(CONF_PORT, DEFAULT_PORT)
+            await get_cert_time_to_expiry(
+                self.hass,
+                user_input[CONF_HOST],
+                user_input.get(CONF_PORT, DEFAULT_PORT),
             )
             return True
-        except socket.gaierror:
-            _LOGGER.error("Host cannot be resolved: %s", host)
-            self._errors[CONF_HOST] = "resolve_failed"
-        except socket.timeout:
-            _LOGGER.error("Timed out connecting to %s", host)
-            self._errors[CONF_HOST] = "connection_timeout"
-        except ssl.CertificateError as err:
-            if "doesn't match" in err.args[0]:
-                _LOGGER.error("Certificate does not match host: %s", host)
-                self._errors[CONF_HOST] = "wrong_host"
-            else:
-                _LOGGER.error("Certificate could not be validated: %s", host)
-                self._errors[CONF_HOST] = "certificate_error"
-        except ssl.SSLError:
-            _LOGGER.error("Certificate could not be validated: %s", host)
-            self._errors[CONF_HOST] = "certificate_error"
+        except TemporaryFailure as err:
+            cause = type(err.__cause__)
```

Comment:
I think this cause resolution belongs in the helper and it should raise a more specific error depending on the cause. That way we put all the socket certificate logic in one place.

A telltale sign is that we still need to import `socket` here in the config flow.

Reply:
Okay to return the error strings from handler?